tree/treeplayer/src/TTreeProcessorMT.cxx

```diff
@@ -37,6 +37,92 @@ struct EntryCluster {
+/// TEntryList-local entry numbers.
+///
+/// This routine assumes that entry numbers in the TEntryList (and, if present, in the sub-entrylists) are in
+/// ascending order, i.e., for n > m:
+///   elist.GetEntry(n) + tree_offset_for_entry_from_elist(n) > elist.GetEntry(m) + tree_offset_for_entry_from_elist(m)
+static std::vector<std::vector<EntryCluster>>
+ConvertToElistClusters(std::vector<std::vector<EntryCluster>> &&clusters, TEntryList &entryList,
+                       const std::vector<std::string> &treeNames, const std::vector<std::string> &fileNames,
+                       const std::vector<Long64_t> &entriesPerFile)
+{
+   R__ASSERT(entryList.GetN() > 0); // wasteful to call this function if it has nothing to do
+   R__ASSERT(ClustersAreSortedAndContiguous(clusters));
+
+   const bool listHasGlobalEntryNumbers = entryList.GetLists() == nullptr;
+   const auto nFiles = clusters.size();
+
+   std::unique_ptr<TChain> chain;
+   using NextFn_t = Long64_t (*)(Long64_t &, TEntryList &, TChain &);
+   // A function that advances TEntryList and returns global entry numbers or -1 if we reached the end
+   // (might or might not need a TChain depending on whether listHasGlobalEntryNumbers)
+   NextFn_t Next;
+   if (listHasGlobalEntryNumbers) {
+      Next = [](Long64_t &elEntry, TEntryList &elist, TChain &) {
+         ++elEntry;
+         return elist.Next();
+      };
+   } else {
+      // we need `chain` to be able to convert local entry numbers to global entry numbers in `Next`
+      chain.reset(new TChain());
+      for (auto i = 0u; i < nFiles; ++i)
+         chain->Add((fileNames[i] + "/" + treeNames[i]).c_str(), entriesPerFile[i]);
+      Next = [](Long64_t &elEntry, TEntryList &elist, TChain &ch) {
+         ++elEntry;
+         int treenum = -1;
+         Long64_t localEntry = elist.GetEntryAndTree(elEntry, treenum);
+         if (localEntry == -1ll)
+            return localEntry;
+         return localEntry + ch.GetTreeOffset()[treenum];
+      };
+   }
+
+   // the call to GetEntry also serves the purpose to reset TEntryList::fLastIndexQueried,
+   // so we can be sure TEntryList::Next will return the correct thing
+   Long64_t elistEntry = 0ll;
+   Long64_t entry = entryList.GetEntry(elistEntry);
+
+   std::vector<std::vector<EntryCluster>> elistClusters;
+
+   for (auto fileN = 0u; fileN < nFiles; ++fileN) {
+      std::vector<EntryCluster> elistClustersForFile;
+      for (const auto &c : clusters[fileN]) {
+         if (entry >= c.end || entry == -1ll)  // no entrylist entries in this cluster
+            continue;
+         R__ASSERT(entry >= c.start);  // current entry should never come before the cluster we are looking at
+         const Long64_t elistRangeStart = elistEntry;
+         // advance entry list until the entrylist entry goes beyond the end of the cluster
+         while (entry < c.end && entry != -1ll)
+            entry = Next(elistEntry, entryList, *chain);
```

Comment:
Isn't that dereferencing a nullptr (unique) pointer when listHasGlobalEntryNumbers is true?

Reply:
Yes but it's doing nothing with it, isn't that safe?

Reply:
ah indeed you are right, already the reference assignment counts as undefined behavior (e.g. https://stackoverflow.com/questions/4364536/is-null-reference-possible). Will fix.

Reply:
fixed